src/lapack/backends/cusolver/cusolver_batch.cpp

```diff
@@ -79,251 +100,645 @@ void getri_batch(sycl::queue &queue, std::int64_t n, sycl::buffer<std::complex<d
+        overflow_check(m[i], n[i], lda[i], group_sizes[i]);
+        batch_size += group_sizes[i];
+    }
+
+    // cuSolver legacy api does not accept 64-bit ints.
+    // To get around the limitation.
+    // Allocate memory with 32-bit ints then copy over results
+    int **ipiv32 = (int **)malloc(sizeof(int *) * batch_size);
+    int64_t global_id = 0;
+    for (int64_t group_id = 0; group_id < group_count; ++group_id)
+        for (int64_t local_id = 0; local_id < group_sizes[group_id]; ++local_id, ++global_id)
+            ipiv32[global_id] = (int *)malloc_device(sizeof(int) * n[group_id], queue);
+    int *devInfo = (int *)malloc_device(sizeof(int) * batch_size, queue);
+
+    auto done = queue.submit([&](sycl::handler &cgh) {
+        int64_t num_events = dependencies.size();
+        for (int64_t i = 0; i < num_events; i++) {
+            cgh.depends_on(dependencies[i]);
+        }
+        onemkl_cusolver_host_task(cgh, queue, [=](CusolverScopedContextHandler &sc) {
+            auto handle = sc.get_handle(queue);
+            auto a_ = reinterpret_cast<cuDataType **>(a);
+            auto scratch_ = reinterpret_cast<cuDataType *>(scratchpad);
+            int64_t global_id = 0;
+            cusolverStatus_t err;
+
+            // Uses scratch so sync between each cuSolver call
+            for (int64_t group_id = 0; group_id < group_count; ++group_id) {
+                for (int64_t local_id = 0; local_id < group_sizes[group_id];
+                     ++local_id, ++global_id) {
+                    CUSOLVER_ERROR_FUNC_T_SYNC(func_name, func, err, handle, m[group_id],
+                                               n[group_id], a_[global_id], lda[group_id], scratch_,
+                                               ipiv32[global_id], devInfo + global_id);
+                }
+            }
+        });
+    });
+
+    // Copy from 32-bit USM to 64-bit
+    std::vector<sycl::event> casting_dependencies(group_count);
+    for (int64_t group_id = 0, global_id = 0; group_id < group_count; ++group_id) {
+        uint64_t ipiv_size = n[group_id];
+        for (int64_t local_id = 0; local_id < group_sizes[group_id]; ++local_id, ++global_id) {
+            int64_t *d_ipiv = ipiv[global_id];
+            int *d_ipiv32 = ipiv32[global_id];
+
+            sycl::event e = queue.submit([&](sycl::handler &cgh) {
+                cgh.depends_on(done);
+                cgh.parallel_for(sycl::range<1>{ ipiv_size }, [=](sycl::id<1> index) {
+                    d_ipiv[index] = static_cast<std::int64_t>(d_ipiv32[index]);
+                });
+            });
+            casting_dependencies[group_id] = e;
+        }
+    }
+
+    // Enqueue free memory
+    sycl::event done_freeing = queue.submit([&](sycl::handler &cgh) {
+        int64_t num_events = casting_dependencies.size();
+        for (int64_t i = 0; i < num_events; i++) {
+            cgh.depends_on(casting_dependencies[i]);
+        }
+        cgh.host_task([=](sycl::interop_handle ih) {
```

Comment:
this would not be required if ipiv was passed as **int32_t, and converted in-pace later,